### DISCUSSION TITLE
Attach VSCode debugger to capacitor app running in Android

### DISCUSSION THREAD

**Discussion by gmaughan** [ionic-team/capacitor#3278] [Q&A] [answered]
I've been running my capacitor app on Android with:

```sh
ionic cap run android -l --external
```

And then either running in an emulator or physical device from Android Studio.

This works well for live reloading. But I've been unable to work out if a VSCode debugger can be attached to set breakpoints etc.

Is this possible? If so, can someone help with the launch.json config to do so?

Thanks.
Glenn.

**Comment by dotNetkow** [comment, marked as answer]
Yes, it is. See the [Ionic docs here](https://ionicframework.com/docs/troubleshooting/debugging#debugging-with-visual-studio-code-in-android).

Capacitor-specific instructions:
0. Ensure your device is connected to the computer.
1. Install the [Android WebView Debugging extension](https://marketplace.visualstudio.com/items?itemName=mpotthoff.vscode-android-webview-debug).
2. In launch.json, configure:

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "android-webview",
            "request": "attach",
            "name": "Attach to Android WebView",
            "webRoot": "${workspaceFolder}/www",
            "sourceMaps": true,
            "sourceMapPathOverrides": {
                "webpack:/*": "${workspaceFolder}/*"
            }
        }
    ]
}
```

3. Start up the app with live reload building, `ionic cap run android -l --external`.
4. Set some breakpoints in Visual Studio Code.
5. From the Debug side menu, click the Play button. A drop down will appear, select the WebView that is running the Cap app.
6. Navigate around the app, etc. to trigger the breakpoint.

**Reply by gmaughan** [reply]
Thank you. Exactly what I was after.

**Reply by jfcere** [reply]
To simplify the _Step 3_ where you start the app with live reload using `ionic cap run android -l --external` you can create a task to do it in `.vscode\tasks.json`.

```jsonc
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "run ionic capacitor android",
      "type": "shell",
      "command": "ionic cap run android -l --external"
    }
  ]
}
```

**Reply by martin-braun** [reply]
@dotNetkow Hi Matt, thanks for this. Could you please tell me how I can attach the debugger on launch automatically to debug init code?

**Comment by ekrapfl** [comment]
Hi, I just wanted to ask if anyone else has any issues getting this to work?  I am running the same commands, and I have the same configuration in my launch.json.  I can launch the app, I can attach the debugger to the correct view, but my breakpoints won't hit.  It seems like a source map problem, but I'm not sure what the issue could be.  I have the same setup as described here.  Where should my sourcemaps be coming from?

One thing that confuses me is that we are pointing to the www directory for the webroot, and then directly to the workspace for the webpack sourcemaps.  I guess I was under the impression that the files are all loaded in memory when you are serving (which the live-reload is essentially doing), or maybe I am way off on that.  Any help would be appreciated.

Thanks!

**Reply by ekrapfl** [reply]
Actually, I think I figured it out.  Not positive why this is the case for me, but I have to NOT set `webRoot` down to the `www` directory, and then it works fine for me.  So my configuration looks like so:
```JSON
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "android-webview",
            "request": "attach",
            "name": "Attach to Android WebView",
            "webRoot": "${workspaceFolder}/*",
            "sourceMaps": true,
            "sourceMapPathOverrides": {
                "webpack:/*": "${workspaceFolder}/*"
            }
        }
    ]
}
```

Fyi, in case anyone else runs into that.

**Reply by harikvpy** [reply]
Thanks for posting this. Good man!

**Reply by harikvpy** [reply]
@dotNetkow I hope you can update the docs for the webRoot configuration as discovered by @ekrapfl

**Reply by dotNetkow** [reply]
Hmm to be honest, I'm not sure I understand what's happening there. What framework are you using? When I originally tested this configuration, it was Ionic Angular.

**Reply by harikvpy** [reply]
Using Ionic/Angular. I too have to set the webRoot to ${workspaceFolder}/* to get it to work.

**Reply by ekrapfl** [reply]
Same here.  I am also using Ionic Angular.  Here is my info output:
```
Ionic:

   Ionic CLI                     : 6.12.3
   Ionic Framework               : @ionic/angular 5.2.3
   @angular-devkit/build-angular : 0.1000.1
   @angular-devkit/schematics    : 10.0.3
   @angular/cli                  : 10.0.1
   @ionic/angular-toolkit        : 2.3.0

Capacitor:

   Capacitor CLI   : 2.4.5
   @capacitor/core : 2.4.5

Utility:

   cordova-res : not installed
   native-run  : 1.3.0

System:

   NodeJS : v12.18.2 (/Users/krapfl/.nvm/versions/node/v12.18.2/bin/node)
   npm    : 6.14.7
   OS     : macOS
```

**Comment by lamieltoch** [comment]
Hi !

I'm working with capacitor vue/vuetify and android with webview. I try to write the above launch.json. Everything work except breakpoints. VSCode tells me that the breakpoints are unbound (greyed on the UI). So the breakpoints are ignored.

How to make breakpoints bound ?

Have you some ideas ?

Thanks

**Reply by Nathan187** [reply]
is anyone else getting 'Failed to locate ADB executable?'  i tried 

For zsh users. Add alias adb='/Users/usrName/Library/Android/sdk/platform-tools/adb' to .zshrc file.

However, it still isn't working.  Any ideas?  (this happens when I tried step 5.

edit....i added the remote-ssh extension for grins.  and now getting a span ENOTDIR error.  any help is appreciated

got it to work but breakpoints are being hit...instead of a red circle , i get a hollow circle when debugger is attached
